Fix update providers script and update aws terraform provider to prove it works

garbas opened the pull request:
I noticed that terraform providers that have unstable release will never update, since the update script will always take the last version and fail if that version is unstable (like it currently does for AWS since last version at the time of writing this is 6.0.0-beta1)

I've changed the update-provide script so that the last stable version is selected. 

Furthermore, I've also run the script to update AWS provider to prove this works.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

garbas committed:
terraform-providers: Only update to latest stable version

garbas committed:
terraform-providers.aws: 5.90.0 -> 5.97.0

zowoq reviewed (dismissed):
This will downgrade providers that are already on an unstable version, e.g. proxmox.

garbas commented:
@zowoq Tnx for a quick look at this PR.

Could a solution be to check:
- if current version is already unstable then take last version even if it unstable
- if current version is stable then take the last stable version

Or do you propose some other solution?

garbas commented:
I do find it odd that there are unstable providers in nixpkgs. Do you maybe know what was the reason for proxmox to be bumped to unstable?

zowoq commented:
> Could a solution be to check: ...

Sure, that sounds fine.

> I do find it odd that there are unstable providers in nixpkgs. Do you maybe know what was the reason for proxmox to be bumped to unstable?

https://github.com/NixOS/nixpkgs/pull/382936 for promox. IIRC there has been a couple of other providers that were basially stuck on pre-release versions like 0.0.1-beta for a quite while or the stable version used an ancient version of go so it wasn't practical to package.

garbas committed:
terraform-providers.proxmox: v3.0.1-rc6 -> v3.0.1-rc8

garbas commented:
@zowoq I've changed the script as we discussed. And I've run it against aws and it correctly reported no change (since I already included the commit to update it) and I've also added the commit for proxmox where you can see that unstable version is preserved.
